packages/arb-avm-cpp/data_storage/include/data_storage/datastorage.hpp

```diff
@@ -56,18 +56,41 @@ class DataStorage {
     size_t next_column_to_flush{0};
     std::vector<uint8_t> secret_hash_seed;
 
+    class shutting_down_exception : public std::exception {};
+
+    class ConcurrentCounter {
+       public:
+        ConcurrentCounter() {
+            if (shutting_down) {
+                throw shutting_down_exception();
```

Comment:
I think if this condition is hit the destructor will still decrement, which would be bad.

Reply:
You might consider initializing the counter via a helper method which does the first shutting_down check, then calls a private constructor, which increments the counter and then does the second shutting_down check.

Reply:
Thanks, that's a lot cleaner